### ISSUE TITLE
Banana negotiation error handling does very poor logging

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2326]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2326|
|Type|defect|
|Created|2006-12-24 18:40:32Z|

Two error paths through dialect negotiation in banana.py:

```python
                else:
                    # I can't speak any of those dialects.
                    log.msg('error losing')
                    self.transport.loseConnection()
            else: 
                if obj in self.knownDialects:
                    self._selectDialect(obj)
                else:
                    # the client just selected a protocol that I did not suggest.
                    log.msg('freaky losing')
                    self.transport.loseConnection()
```

Some more informative messages would be useful there.

Attachments:

* [banana-logging.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d1c/d1c2fc6828b03a590a5298569f9bd616af775020/46d820e44268763947f94f84a916fa12b7b1f7af.patch) (1047 bytes) - added by jesstess on 2009-08-31 18:43:48Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2326 2326
type__defect defect
reporter__exarkun exarkun
priority__low low
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__pb pb
keywords__ 
time__1166985632000000 1166985632000000
changetime__1251744940000000 1251744940000000
version__None None
owner__ 
cc__thijs cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_1"></a>|@jesstess removed owner|
|-|-|

How do "The client doesn't speak any of the dialects offered by the server; disconnecting." and "The client selected a dialect the server didn't suggest and doesn't know; disconnecting." sound?

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_2"></a>|@thijstriemstra set owner to @jesstess|
|-|-|

I don't like the use of the ```;``` too much, at first it looked like a typo to me. Also not sure about the word ```dialect```, what about ```protocol``` and a ```:``` instead?

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_3"></a>|@jesstess removed owner|
|-|-|

Patch changed to "The client doesn't speak any of the protocols offered by the server: disconnecting." and "The client selected a protocol the server didn't suggest and doesn't know: disconnecting."

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_4"></a>|@thijstriemstra set owner to @thijstriemstra<br>@thijstriemstra set status to `assigned`|
|-|-|

Definitely an improvement over the current ```log.msg```, applying patch.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra set status to `closed`|
|-|-|

(In [27268]) Apply banana-logging.patch: Improve the negotiation log messages.

Author: jesstess
Reviewer: thijs
Fixes: [#2326](https://github.com/twisted/twisted/issues/2326)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
